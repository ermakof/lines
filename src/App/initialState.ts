import { IAppState } from '@src/App/model/IAppState';
import createGameField from '@src/utils/createGameField';

export const GAME_FIELD_SIZE = 9;

export const INIT_GAME_LEVEL = '1';
export const INIT_GAME_FIELD_PERCENT_FILLED = 0;

interface ISettings {
  percent: number;
  chain: number;
  increase: number;
}
interface IProps {
  [key: string]: ISettings;
}
export const GAME_LEVEL_SETTINGS: IProps = {
  '1': { percent: 10, chain: 2, increase: 3 },
  '2': { percent: 20, chain: 3, increase: 4 },
  '3': { percent: 30, chain: 4, increase: 5 },
};

const initialState: IAppState = {
  userLevel: INIT_GAME_LEVEL,
  gameFieldPercentFilled: 0,
  gameFieldData: createGameField(GAME_FIELD_SIZE ** 2, INIT_GAME_FIELD_PERCENT_FILLED),
  highlightedCells: undefined,
  score: 0,
  selectedCell: undefined,
  isLoading: false,
  hitParade: undefined,
};

export default initialState;
