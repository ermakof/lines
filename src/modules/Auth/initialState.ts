import { IAuthState } from '@src/modules/Auth/model/IAuthState';

const initialState: IAuthState = {
  userProfile: undefined,
};

export default initialState;
