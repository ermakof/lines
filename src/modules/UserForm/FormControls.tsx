/* eslint-disable react-hooks/exhaustive-deps */
import React, { FC, memo } from 'react';
import { useDispatch } from 'react-redux';

import Button from '@src/components/Button';
import { actions as appActions } from '@src/App/appSlice';
import { actions as authActions } from '@src/modules/Auth/authSlice';
import Select from '@src/components/Select';
import styled from '@emotion/styled';

const Root = styled.div`
  display: flex;
`;

interface IFormControls {
  userLevel?: string;
}
const FormControls: FC<IFormControls> = ({ userLevel }) => {
  const dispatch = useDispatch();

  const handleReset = () => {
    dispatch(appActions.resetApp());
  };

  const handleSelectLevel = (event: React.ChangeEvent<HTMLSelectElement>) => {
    const { value } = event.currentTarget;
    dispatch(appActions.setUserLevel(value));
  };

  const handleLogout = () => {
    dispatch(authActions.logout());
  };

  return (
    <Root>
      <Button role="buttonReset" onClick={handleReset} title="Сброс" />
      <Select onSelect={handleSelectLevel} selectedLevel={userLevel} />
      <Button role="buttonLogout" onClick={handleLogout} title="Выйти" />
    </Root>
  );
};

export default memo(FormControls);
