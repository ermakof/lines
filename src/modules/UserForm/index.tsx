/* eslint-disable react-hooks/exhaustive-deps */
import React, { FC, memo, useEffect } from 'react';
import { useDispatch } from 'react-redux';

import Panel from '@src/layout/Panel';
import Button from '@src/components/Button';
import { actions as appActions } from '@src/App/appSlice';
import { actions as authActions } from '@src/modules/Auth/authSlice';
import Select from '@src/components/Select';

const UserForm: FC = () => {
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(appActions.rehydrate());
  }, []);

  const handleReset = () => {
    dispatch(appActions.resetApp());
  };

  const handleSelectLevel = (event: React.ChangeEvent<HTMLSelectElement>) => {
    const { value } = event.currentTarget;
    dispatch(appActions.setUserLevel(value));
  };

  const handleLogout = () => {
    dispatch(authActions.logout());
  };

  return (
    <Panel role="userForm">
      <Button role="buttonReset" onClick={handleReset} title="Reset" />
      <Select onSelect={handleSelectLevel} />
      <Button role="buttonLogout" onClick={handleLogout} title="Выйти" />
    </Panel>
  );
};

export default memo(UserForm);
