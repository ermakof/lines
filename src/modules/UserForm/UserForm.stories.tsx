import React from 'react';
import { Provider } from 'react-redux';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withKnobs } from '@storybook/addon-knobs';
import { BrowserRouter as Router } from 'react-router-dom';

import UserForm from '@src/modules/UserForm';
import { store } from '@src/store';

const dispatch = () => null;

const state = {
  gameLevel: '1',
  gameFieldSize: 4,
  gameFieldPercentFilled: 10,
  gameFieldData: [0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0],
};

export default {
  component: UserForm,
  decorators: [withKnobs],
  title: 'Forms/UserForm',
} as ComponentMeta<typeof UserForm>;

const Template: ComponentStory<typeof UserForm> = (args) => (
  <Provider store={store}>
    <Router>
      <UserForm {...args} />
    </Router>
  </Provider>
);
export const Static = Template.bind({});

Static.args = {
  onSubmit: action('submitted'),
};
