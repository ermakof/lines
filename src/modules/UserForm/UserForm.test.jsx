import React from 'react';
import { Provider, useDispatch, useSelector } from 'react-redux';
import { render, screen, waitFor } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import configureMockStore from 'redux-mock-store';

import UserForm from '@src/modules/UserForm';
import initialState from '@src/App/initialState';

const mockStore = configureMockStore();

jest.mock('react-redux', () => ({
  __esModule: true,
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

let store;
let mockDispatch;

describe('UserForm', () => {
  beforeEach(() => {
    store = mockStore(initialState);
    mockDispatch = jest.fn();
    useDispatch.mockReturnValue(mockDispatch);
    useSelector.mockReturnValue(initialState);
  });

  it('Render <UserForm>', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Router>
          <UserForm />
        </Router>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
    const userForm = screen.getByRole(/userForm/gi);
    expect(userForm).toBeInTheDocument();
    const buttonReset = screen.getByRole(/^buttonReset$/gi);
    expect(buttonReset).toBeInTheDocument();
    const select = screen.getByRole(/^select$/gi);
    expect(select).toBeInTheDocument();
    const buttonLogout = screen.getByRole(/^buttonLogout$/gi);
    expect(buttonLogout).toBeInTheDocument();
  });

  it('Reset game <UserForm>', async () => {
    render(
      <Provider store={store}>
        <Router>
          <UserForm />
        </Router>
      </Provider>
    );
    userEvent.click(screen.getByRole('buttonReset'));

    await waitFor(() =>
      expect(mockDispatch).toHaveBeenCalledWith({
        type: 'app/resetApp',
      })
    );
  });

  it('Logout app <UserForm>', async () => {
    render(
      <Provider store={store}>
        <Router>
          <UserForm />
        </Router>
      </Provider>
    );
    userEvent.click(screen.getByRole('buttonLogout'));

    await waitFor(() =>
      expect(mockDispatch).toHaveBeenCalledWith({
        type: 'auth/logout',
      })
    );
  });

  it('Select level <UserForm>', async () => {
    render(
      <Provider store={store}>
        <Router>
          <UserForm />
        </Router>
      </Provider>
    );
    const select = screen.getByRole(/select/gi);
    userEvent.selectOptions(select, '2');

    await waitFor(() =>
      expect(mockDispatch).toHaveBeenCalledWith({
        type: 'app/setUserLevel',
        payload: '2',
      })
    );
  });
});
