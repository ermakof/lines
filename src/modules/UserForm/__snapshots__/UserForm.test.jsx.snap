// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserForm Render <UserForm> 1`] = `
<DocumentFragment>
  <div
    class="css-lgrtma"
    role="userForm"
  >
    <button
      class="css-tv5zn1"
      role="buttonReset"
    >
      Reset
    </button>
    <select
      class="css-tv5zn1"
      name="select"
      role="select"
    >
      <option
        role="option"
        value="1"
      >
        Падаван
      </option>
      <option
        role="option"
        value="2"
      >
        Джедай
      </option>
      <option
        role="option"
        value="3"
      >
        Мастер-джедай
      </option>
      <option
        role="option"
        value="4"
      >
        Член Совета Силы
      </option>
    </select>
    <button
      class="css-tv5zn1"
      role="buttonLogout"
    >
      Выйти
    </button>
  </div>
</DocumentFragment>
`;
